corehq/apps/reports/standard/sms.py

```diff
@@ -1103,4 +1109,177 @@ class PhoneNumberReport(BaseCommConnectLogReport):
+
+    def _get_queryset(self):
+        query = PhoneNumber.objects.filter(domain=self.domain)
+
+        if self.filter_type == 'phone_number':
+            if self.phone_number_filter:
+                query = query.filter(phone_number__contains=self.phone_number_filter)
+        elif self.filter_type == 'contact':
+            if self._show_cases:
+                query = query.filter(owner_doc_type='CommCareCase', phone_number__isnull=False)
+            else:
+                query = query.filter(owner_doc_type__in=['CommCareUser', 'WebUser'])
+                if self.selected_group:
+                    query = query.filter(owner_id__in=self.user_ids_in_selected_group)
+
+                if self.has_phone_number == 'has_phone_number':
+                    query = query.filter(phone_number__isnull=False)
+
+            if self.verification_status == 'not_verified':
+                query = query.filter(pending_verification=False, verified=False)
+            elif self.verification_status == 'verification_pending':
+                query = query.filter(pending_verification=True)
+            elif self.verification_status == 'verified':
+                query = query.filter(verified=True)
+
+        return query.order_by('phone_number', 'couch_id')
+
+    def _get_users_without_phone_numbers(self):
+        query = (
+            PhoneNumber.objects.filter(domain=self.domain).filter(owner_doc_type__in=['CommCareUser', 'WebUser'])
+        )
+        users_with_phone_numbers = {x[0]: x for x in query.values_list('owner_id', 'owner_doc_type').distinct()}
+
+        if self.selected_group:
+            users_by_id = {
+                id: {'_id': id, 'doc_type': 'CommCareUser'}
+                for id in self.user_ids_in_selected_group
+            }
+        else:
+            users_by_id = {u.get_id: u for u in CouchUser.by_domain(self.domain)}
+
+        user_ids = set(users_by_id.keys()) - set(users_with_phone_numbers.keys())
+        user_types_with_id = {(id, users_by_id[id]['doc_type']) for id in user_ids}
+
+        FakePhoneNumber = namedtuple('FakePhoneNumber', ['owner_id', 'owner_doc_type'])
+        return sorted([FakePhoneNumber(id, type) for id, type in user_types_with_id])
+
+    @property
+    def shared_pagination_GET_params(self):
+        return [
+            {'name': 'filter_type', 'value': self.filter_type},
+            {'name': 'phone_number_filter', 'value': self.phone_number_filter},
+            {'name': 'contact_type', 'value': self.contact_type},
+            {'name': 'selected_group', 'value': self.selected_group},
+            {'name': 'has_phone_number', 'value': self.has_phone_number},
+            {'name': 'verification_status', 'value': self.verification_status},
+        ]
+
+    @property
+    def rows(self):
+        return self._get_rows()
+
+    @property
+    def total_records(self):
+        if self._show_users_without_phone_numbers:
+            return len(self._get_users_without_phone_numbers())
```

Comment:
i think we can memoize `_get_users_without_phone_numbers()` so it doesn't get called twice